File: backend/src/main/java/com/paf/chop/backend/services/UserService.java
// src/main/java/com/paf/chop/backend/services/UserService.java
package com.paf.chop.backend.services;

import com.paf.chop.backend.dto.response.UserResponseDTO;
import com.paf.chop.backend.dto.response.user.PublicUserResponseDTO;
import com.paf.chop.backend.enums.CoinType;
import com.paf.chop.backend.enums.NotificationType;
import com.paf.chop.backend.models.User;
import com.paf.chop.backend.repositories.UserRepository;
import com.paf.chop.backend.utils.ApiResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final NotificationService notificationService;

    /**
     * Resolve an authenticated email to its User ID.
     */
    public Long findIdByUsername(String username) {
        User user = userRepository.findByUsername(username);
        if (user == null) {
            throw new IllegalArgumentException("User not found: " + username);
        }
        return user.getId();
    }



    /**
     * Build full profile DTO for /me.
     */
    @Transactional(readOnly = true)
    public UserResponseDTO buildFullProfile(Long userId) {
        User u = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userId));

        UserResponseDTO dto = new UserResponseDTO(u);
        dto.setFollowerCount(u.getFollowers().size());
        dto.setFollowingCount(u.getFollowing().size());

        dto.setFollowers(
                u.getFollowers().stream()
                        .map(PublicUserResponseDTO::new)
                        .collect(Collectors.toList())
        );
        dto.setFollowing(
                u.getFollowing().stream()
                        .map(PublicUserResponseDTO::new)
                        .collect(Collectors.toList())
        );
        return dto;
    }

    /**
     * Toggle follow/unfollow between two user IDs.
     */
    @Transactional
    public String toggleFollowByIds(Long currentUserId, Long targetUserId) {
        if (currentUserId.equals(targetUserId)) {
            return "Cannot follow yourself";
        }
        User current = userRepository.findById(currentUserId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + currentUserId));
        User target = userRepository.findById(targetUserId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + targetUserId));

        if (current.getFollowing().remove(target)) {
            target.getFollowers().remove(current);
            userRepository.save(current);
            userRepository.save(target);
            return "Unfollowed successfully";
        } else {
            current.getFollowing().add(target);
            target.getFollowers().add(current);
            // optional coin/verification logic
            current.setCoins(current.getCoins() + 10);
            if (current.getCoins() >= 100) current.setIsVerified(true);

            userRepository.save(current);
            userRepository.save(target);

            notificationService.createNotification(
                    current.getUsername() + " started following you",
                    NotificationType.FOLLOW,
                    target.getId()
            );

            addUserCoins(target.getId(), CoinType.FOLLOW);

            return "Followed successfully";
        }
    }

    /**
     * Return public DTO list of followers.
     */
    @Transactional(readOnly = true)
    public List<PublicUserResponseDTO> getFollowersDto(Long userId) {
        User u = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userId));
        return u.getFollowers().stream()
                .map(PublicUserResponseDTO::new)
                .collect(Collectors.toList());
    }

    /**
     * Return public DTO list of following.
     */
    @Transactional(readOnly = true)
    public List<PublicUserResponseDTO> getFollowingDto(Long userId) {
        User u = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userId));
        return u.getFollowing().stream()
                .map(PublicUserResponseDTO::new)
                .collect(Collectors.toList());
    }

    /**
     * Update profile fields for /me.
     */
    @Transactional
    public void updateProfile(Long userId, Map<String, String> updates) {
        User u = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found: " + userId));

        if (updates.containsKey("username")) u.setUsername(updates.get("username"));
        if (updates.containsKey("bio")) u.setBio(updates.get("bio"));
        if (updates.containsKey("profileImage")) u.setProfileImage(updates.get("profileImage"));
        u.setUpdatedAt(LocalDateTime.now());

        userRepository.save(u);
    }

    /**
     * Delete a user by ID.
     */
    @Transactional
    public void deleteUser(Long userId) {
        userRepository.deleteById(userId);
    }

    /**
     * List all users as public DTOs.
     */
    @Transactional(readOnly = true)
    public ApiResponse<List<PublicUserResponseDTO>> getAllUsers() {
        List<User> users = userRepository.findAll();
        List<PublicUserResponseDTO> dtos = users.stream()
                .map(PublicUserResponseDTO::new)
                .collect(Collectors.toList());
        return ApiResponse.success(dtos, "Users fetched");
    }

    /**
     * Get raw User entity (if ever needed).
     */
    @Transactional(readOnly = true)
    public java.util.Optional<User> getUser(Long id) {
        return userRepository.findById(id);
    }


    public Boolean isUserExists(Long userId) {
        return userRepository.existsById(userId);
    }

    public User getCurrentUser() {
        String currentUsername = SecurityContextHolder.getContext().getAuthentication().getName();
        return userRepository.findByUsername(currentUsername);
    }

    /**
     * Coins are added to the user based on the CoinType.
     */
    public void addUserCoins(Long userId, CoinType coinType) {
        try {
            User user = userRepository.findById(userId).orElseThrow(
                    () -> new RuntimeException("User not found")
            );

            user.setCoins(user.getCoins() + coinType.getPoints());
            userRepository.save(user);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public ApiResponse<Integer> getMyCoins() {
        try {
            User user = getCurrentUser();
            return ApiResponse.success(user.getCoins(), "Coins fetched");
        } catch (Exception e) {
            return ApiResponse.error("Error fetching coins: " + e.getMessage());
        }
    }

}

File: frontend/src/services/user.service.ts
// src/services/userService.ts
import api from "@/utils/axiosConfig.ts";

//
//  Generic API wrapper
//
export interface ApiResponse<T> {
  success: boolean;
  message: string;
  result: T;
}

//
//  Public user view (matches PublicUserResponseDTO)
//
export interface PublicUser {
  id: number;
  username: string;
  firstName: string;
  lastName: string;
  profileImage: string;
  bio: string;
}

//
//  Full current‐user view (matches UserResponseDTO)
//
export interface UserProfile extends PublicUser {
  email: string;
  isVerified: boolean;
  coins: number;
  totalLikes: number;
  totalPost: number;
  userRole: string;
  followerCount: number;
  followingCount: number;
  followers: PublicUser[];
  following: PublicUser[];
}

//
//  Endpoints
//

/**
 * GET /api/users/me
 * Returns the full profile of the logged-in user.
 */
export const getCurrentUser = async (): Promise<UserProfile | null> => {
  try {
    const resp = await api.get<UserProfile>("/users/me");
    return resp.data;
  } catch (err) {
    console.error("getCurrentUser error:", err);
    return null;
  }
};

/**
 * GET /api/users/{id}
 * Returns the raw User entity (if you still need it).
 */
export const getUserById = async (id: number): Promise<UserProfile | null> => {
  try {
    const resp = await api.get<UserProfile>(`/users/${id}`);
    return resp.data;
  } catch (err) {
    console.error(`getUserById(${id}) error:`, err);
    return null;
  }
};

/**
 * POST /api/users/{currentUserId}/follow/{targetUserId}
 * Toggle follow/unfollow; returns a success message.
 */
export const toggleFollow = async (
  currentUserId: number,
  targetUserId: number,
): Promise<string | null> => {
  try {
    const resp = await api.post<ApiResponse<string>>(
      `/users/${currentUserId}/follow/${targetUserId}`,
    );
    return resp.data.result;
  } catch (err) {
    console.error("toggleFollow error:", err);
    return null;
  }
};

/**
 * GET /api/users/{id}/followers
 * List of public profiles following user {id}.
 */
export const getFollowers = async (id: number): Promise<PublicUser[]> => {
  try {
    const resp = await api.get<PublicUser[]>(`/users/${id}/followers`);
    return resp.data;
  } catch (err) {
    console.error("getFollowers error:", err);
    return [];
  }
};

/**
 * GET /api/users/{id}/following
 * List of public profiles user {id} is following.
 */
export const getFollowing = async (id: number): Promise<PublicUser[]> => {
  try {
    const resp = await api.get<PublicUser[]>(`/users/${id}/following`);
    return resp.data;
  } catch (err) {
    console.error("getFollowing error:", err);
    return [];
  }
};

/**
 * POST /api/users/upload
 * Upload a profile picture; returns the URL path.
 */
export const uploadProfileImage = async (
  file: File,
): Promise<string | null> => {
  const form = new FormData();
  form.append("file", file);
  try {
    const resp = await api.post<string>("/users/upload", form, {
      headers: { "Content-Type": "multipart/form-data" },
    });
    return resp.data;
  } catch (err) {
    console.error("uploadProfileImage error:", err);
    return null;
  }
};

/**
 * PUT /api/users/me
 * Update the logged-in user’s profile.
 * @param updates partial map of fields you want to change
 */
export interface ProfileUpdates {
  username?: string;
  bio?: string;
  profileImage?: string;
}

export const updateProfile = async (
  updates: ProfileUpdates,
): Promise<boolean> => {
  try {
    const resp = await api.put<ApiResponse<boolean>>("/users/me", updates);
    return resp.data.success;
  } catch (err) {
    console.error("updateProfile error:", err);
    return false;
  }
};

/**
 * DELETE /api/users/me
 * Delete the logged-in user’s account.
 */
export const deleteCurrentUser = async (): Promise<boolean> => {
  try {
    const resp = await api.delete<ApiResponse<boolean>>("/users/me");
    return resp.data.success;
  } catch (err) {
    console.error("deleteCurrentUser error:", err);
    return false;
  }
};

/**
 * GET /api/users/all
 * Fetch all users as public profiles.
 */
export const getAllUsers = async (): Promise<PublicUser[]> => {
  try {
    const resp = await api.get<ApiResponse<PublicUser[]>>("/users/all");
    return resp.data.result;
  } catch (err) {
    console.error("getAllUsers error:", err);
    return [];
  }
};

//call this
export const getUserCoins = async (): Promise<number> => {
  try{
    const resp = await api.get<ApiResponse<number>>("/users/my-coins");
    return resp.data.result;

  }catch (e) {
    console.error("getUserCoins error:", e);
    return 0;
  }
}
